rest-openehr/src/main/java/org/ehrbase/rest/openehr/OpenehrDefinitionQueryController.java

```diff
@@ -121,22 +128,51 @@ public ResponseEntity<QueryDefinitionResponseData> putStoreQuery(
         log.debug("putStoreQuery invoked with the following input: " + qualifiedQueryName + ", version:" + version
                 + ", query:" + queryPayload + ", type=" + type);
 
-        // use the payload from adhoc POST:
-        // get the query and parameters if any
-        Gson gson = new GsonBuilder().create();
-
-        Map<String, Object> mapped = gson.fromJson(queryPayload, Map.class);
-        String aql = (String) mapped.get("q");
+        var format = extractCompositionFormat(contentType);
+        String aql;
+        switch (format) {
+            case JSON: {
+                // use the payload from adhoc POST:
+                // get the query and parameters if any
+                Gson gson = new GsonBuilder().create();
+                Map<String, Object> mapped = gson.fromJson(queryPayload, Map.class);
+                aql = (String) mapped.get("q");
+                break;
+            }
+            case TEXT: {
+                aql = queryPayload;
+                break;
+            }
+            default:
+                throw new UnexpectedSwitchCaseException(format);
+        }
 
         if (aql == null || aql.isEmpty())
             return new ResponseEntity(
                     new ErrorBodyPayload("Invalid query", "no aql query provided in payload").toString(),
                     HttpStatus.BAD_REQUEST);
 
-        QueryDefinitionResponseData queryDefinitionResponseData = new QueryDefinitionResponseData(
-                queryService.createStoredQuery(qualifiedQueryName, version.orElse(null), aql));
+        return internalPutDefinitionProcessing(qualifiedQueryName, version, format, aql);
+    }
 
-        return ResponseEntity.ok(queryDefinitionResponseData);
+    private ResponseEntity<QueryDefinitionResponseData> internalPutDefinitionProcessing(
```

Comment:
The name seems a bit arbitraty;
Maybe we should keep queryService::createStoredQuery in the caller and only prepare the response here.

Reply:
changed